### DISCUSSION TITLE
Connecting Directly to the On-Device ADB Daemon

### DISCUSSION THREAD

**Discussion by JohnPeng47** [yume-chan/ya-webadb#560] [Q&A]
So, actually not a feature request exactly. But I am looking to use your project in another project of my own which is to implement a Android virtual controller application. Most android games works with mobile gaming controllers but for certain games, this feature is not supported (Genshin, looking at you 0__0). Therefore, I need to implement a virtual controller that can take my game controller inputs and generate touch/swipe gestures at the coordinates of the on screen buttons (for the game). Something like this:
https://www.youtube.com/watch?v=Je0UoByNnmM

I cant use INJECT_EVENT permissions because its not a systems app. So alternatively, I'm exploring possibility of running an adb server as an Android service that my virtual controller app can connect to and issue input commands directly via "adb shell input <x> <y>". 

Your code base is a blessing for me as it implements the message passing between the adb server and the adb daemon. However, one requirement that I have which is not implemented directly by you is I need to connect directly to the daemon running on the device itself, through the device loopback rather than through the ADB USB/Wifi protocol.

Do you have any idea how this could be achieved? Apologies for spamming this as a features request but I wasn't sure how else to reach you. I await your reply with great anticipation

**Comment by yume-chan** [comment]
> through the device loopback

I don't think Android's USB host has a loopback.

> connect directly to the daemon running on the device itself

ADB over WiFi (requires a wired connection to turn on after each reboot) and ADB pair mode (on Android 11+) should work.

The pair mode works like this:

1. Enable wireless debugging in developer settings
2. (Only the first time) Pair with the device using a password generated by device (let device trust the client's key). This process is not implemented in this library
3. Connect to device same as ADB over WiFi
